### PR TITLE
fix: missing `__getitem__` type fixes

### PR DESCRIPTION
**Pull request by EdAbati**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [x] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below

I am so sorry @MarcoGorelli , I was a bit to quick to press "Ready for Review" 😭

This is a follow up of #1958

### CODE REVIEW

File: narwhals/dataframe.py [hunk 1]
```diff
@@ -858,16 +858,18 @@ def estimated_size(self: Self, unit: SizeUnit = "b") -> int | float:
...
-        self: Self, key: str | tuple[slice | Sequence[int] | np.ndarray, int | str]
+        self: Self,
+        item: str | tuple[slice | Sequence[int] | np.ndarray, int | str],
     ) -> Series[Any]: ...
 
     @overload
     def __getitem__(
         self: Self,
-        key: (
+        item: (
```
Comment (EdAbati) [inline narwhals/dataframe.py:868]: silly mistake (I blame copilot) 🙈

I wonder why it was not picked up though

File: narwhals/dataframe.py [hunk 2]
```diff
@@ -858,16 +858,18 @@ def estimated_size(self: Self, unit: SizeUnit = "b") -> int | float:
...
             slice
             | Sequence[int]
             | Sequence[str]
+            | np.ndarray
```
Comment (EdAbati) [inline narwhals/dataframe.py:872]: this was missing too

File: tests/frame/get_column_test.py [hunk 3]
```diff
@@ -29,5 +29,5 @@ def test_non_string_name() -> None:
 
 def test_get_single_row() -> None:
     df = pd.DataFrame({"a": [1, 2], "b": [3, 4]})
-    result = nw.from_native(df, eager_only=True)[0]  # type: ignore[call-overload]
+    result = nw.from_native(df, eager_only=True)[0]
```
Comment (EdAbati) [inline tests/frame/get_column_test.py:32]: these are making `mypy` happy probably because we have `def __getitem__(self: Self, item: Any) -> Any:` in `narwhals.stable.v1`

Pyright is still not happy 

We really should stabilise the `mypy` checks 🥲

**Comment by dangotbanned** [comment]
fix(typing): fix some of `DataFrame.__getitem__`
```diff
diff --git a/narwhals/dataframe.py b/narwhals/dataframe.py
index 41c8edff..dff182a9 100644
--- a/narwhals/dataframe.py
+++ b/narwhals/dataframe.py
@@ -858,8 +858,7 @@ class DataFrame(BaseFrame[DataFrameT]):
 
     @overload
     def __getitem__(  # type: ignore[overload-overlap]
-        self: Self,
-        item: str | tuple[slice | Sequence[int] | np.ndarray, int | str],
+        self: Self, item: str | tuple[slice | Sequence[int] | np.ndarray, int | str]
     ) -> Series[Any]: ...
 
     @overload
@@ -869,10 +868,7 @@ class DataFrame(BaseFrame[DataFrameT]):
             slice
             | Sequence[int]
             | Sequence[str]
-            | np.ndarray
-            | tuple[
-                slice | Sequence[int] | np.ndarray, slice | Sequence[int] | Sequence[str]
-            ]
+            | tuple[slice | np.ndarray, slice | Sequence[int] | Sequence[str]]
         ),
     ) -> Self: ...
     def __getitem__(
@@ -882,11 +878,8 @@ class DataFrame(BaseFrame[DataFrameT]):
             | slice
             | Sequence[int]
             | Sequence[str]
-            | np.ndarray
             | tuple[slice | Sequence[int] | np.ndarray, int | str]
-            | tuple[
-                slice | Sequence[int] | np.ndarray, slice | Sequence[int] | Sequence[str]
-            ]
+            | tuple[slice | np.ndarray, slice | Sequence[int] | Sequence[str]]
         ),
     ) -> Series[Any] | Self:
         """Extract column or slice of DataFrame.
diff --git a/narwhals/stable/v1/__init__.py b/narwhals/stable/v1/__init__.py
index 25218a09..aef4f27c 100644
--- a/narwhals/stable/v1/__init__.py
+++ b/narwhals/stable/v1/__init__.py
@@ -144,9 +144,7 @@ class DataFrame(NwDataFrame[IntoDataFrameT]):
             | np.ndarray
             | Sequence[int]
             | Sequence[str]
-            | tuple[
-                slice | Sequence[int] | np.ndarray, slice | Sequence[int] | Sequence[str]
-            ]
+            | tuple[slice | np.ndarray, slice | Sequence[int] | Sequence[str]]
         ),
     ) -> Self: ...
 

```

**Comment by MarcoGorelli** [comment]
> I haven't got permissions

you do now 😉 feel free to make improvements, thanks for all your help + evangelism of Narwhals

File: tests/frame/get_column_test.py (hunk 3 above)
Comment (dangotbanned) [reply tests/frame/get_column_test.py:32]: Resolved in (https://github.com/narwhals-dev/narwhals/pull/1963/commits/932a2f93deb2d83960ea43cee3d5acd3caa9081d)

Comment (dangotbanned) [reply tests/frame/get_column_test.py:32]: CI failure is a false-positive (https://results.pre-commit.ci/run/github/760058710/1739017689.-NTAkR4uTPqdmica09T0cg)

```py
>>> mypy tests/frame/getitem_test.py
tests\frame\getitem_test.py:66: error: No overload variant of "__getitem__" of "DataFrame" matches argument type "Foo"  [call-overload]
tests\frame\getitem_test.py:66: note: Possible overload variants:
tests\frame\getitem_test.py:66: note:     def __getitem__(self, str | tuple[slice[Any, Any, Any] | Sequence[int] | ndarray[Any, Any], int | str], /) -> Series
tests\frame\getitem_test.py:66: note:     def __getitem__(self, int | slice[Any, Any, Any] | ndarray[Any, Any] | Sequence[int] | Sequence[str] | tuple[slice[Any, Any, Any] | Sequence[int] | ndarray[Any, Any], slice[Any, Any, Any] | Sequence[int] | Sequence[str]], /) -> narwhals.stable.v1.DataFrame[NativeFrame | narwhals.dataframe.DataFrame[Any] | DataFrameLike]
Found 1 error in 1 file (checked 1 source file)
```

Getting two warnings from `pyright`:

```
No overloads for "__getitem__" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
Argument of type "Foo" cannot be assigned to parameter "item" of type "int | slice[Any, Any, Any] | ndarray[Unknown, Unknown] | Sequence[int] | Sequence[str] | tuple[slice[Any, Any, Any] | Sequence[int] | ndarray[Unknown, Unknown], slice[Any, Any, Any] | Sequence[int] | Sequence[str]]" in function "__getitem__"
  Type "Foo" is not assignable to type "int | slice[Any, Any, Any] | ndarray[Unknown, Unknown] | Sequence[int] | Sequence[str] | tuple[slice[Any, Any, Any] | Sequence[int] | ndarray[Unknown, Unknown], slice[Any, Any, Any] | Sequence[int] | Sequence[str]]"
    "Foo" is not assignable to "int"
    "Foo" is not assignable to "slice[Any, Any, Any]"
    "Foo" is not assignable to "ndarray[Unknown, Unknown]"
    "Foo" is not assignable to "Sequence[int]"
    "Foo" is not assignable to "Sequence[str]"
    "Foo" is not assignable to "tuple[slice[Any, Any, Any] | Sequence[int] | ndarray[Unknown, Unknown], slice[Any, Any, Any] | Sequence[int] | Sequence[str]]"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

**Comment by dangotbanned** [comment]
Maybe this is too much for the PR.
But from someone who doesn't use `__getitem__` - replacing these types with the [`polars` aliases](https://github.com/pola-rs/polars/blob/ff994f0502bc5ce886df43ffbbf468a79adcf281/py-polars/polars/_typing.py#L281-L303) would be much easier to read IMO.

```py
# Annotations for `__getitem__` methods
SingleIndexSelector: TypeAlias = int
MultiIndexSelector: TypeAlias = Union[
    slice,
    range,
    Sequence[int],
    "Series",
    "np.ndarray[Any, Any]",
]
SingleNameSelector: TypeAlias = str
MultiNameSelector: TypeAlias = Union[
    slice,
    Sequence[str],
    "Series",
    "np.ndarray[Any, Any]",
]
BooleanMask: TypeAlias = Union[
    Sequence[bool],
    "Series",
    "np.ndarray[Any, Any]",
]
SingleColSelector: TypeAlias = Union[SingleIndexSelector, SingleNameSelector]
MultiColSelector: TypeAlias = Union[MultiIndexSelector, MultiNameSelector, BooleanMask]
```

Still ends up pretty complex [in the `@overload`(s)](https://github.com/pola-rs/polars/blob/ff994f0502bc5ce886df43ffbbf468a79adcf281/py-polars/polars/dataframe/frame.py#L1199-L1235), but much easier to differentiate the parts:

```py
    # `str` overlaps with `Sequence[str]`
    # We can ignore this but we must keep this overload ordering
    @overload
    def __getitem__(
        self, key: tuple[SingleIndexSelector, SingleColSelector]
    ) -> Any: ...

    @overload
    def __getitem__(  # type: ignore[overload-overlap]
        self, key: str | tuple[MultiIndexSelector, SingleColSelector]
    ) -> Series: ...

    @overload
    def __getitem__(
        self,
        key: (
            SingleIndexSelector
            | MultiIndexSelector
            | MultiColSelector
            | tuple[SingleIndexSelector, MultiColSelector]
            | tuple[MultiIndexSelector, MultiColSelector]
        ),
    ) -> DataFrame: ...

    def __getitem__(
        self,
        key: (
            SingleIndexSelector
            | SingleColSelector
            | MultiColSelector
            | MultiIndexSelector
            | tuple[SingleIndexSelector, SingleColSelector]
            | tuple[SingleIndexSelector, MultiColSelector]
            | tuple[MultiIndexSelector, SingleColSelector]
            | tuple[MultiIndexSelector, MultiColSelector]
        ),
    ) -> DataFrame | Series | Any:
```

@MarcoGorelli do you see these aliases as a desirable thing to steal from `polars`?

File: narwhals/dataframe.py [hunk 4]
```diff
@@ -861,16 +861,19 @@ def estimated_size(self: Self, unit: SizeUnit = "b") -> int | float:
...
+        item: str | tuple[slice | Sequence[int] | np.ndarray, int | str],
     ) -> Series[Any]: ...
 
     @overload
     def __getitem__(
         self: Self,
-        key: (
-            slice
+        item: (
+            int
```
Comment (EdAbati) [inline narwhals/dataframe.py:872]: Hey @dangotbanned thanks for fixing! super quick I didn't have the chance to see the comments :D

Anyway I think that `int` is a bit of a special case.
The narwhals rule at the moment is that `int` in `__getitem__` should not be fully supported (https://narwhals-dev.github.io/narwhals/pandas_like_concepts/column_names/) 
That will work in pandas but not everywhere.

I think we shouldn't add it in typing here, but ignore the `int` slicing in the tests if a typing error occurs.

Having said that, it is weird that after my changes `mypy` wants me to remove the `# ignore` in the tests

**Comment by EdAbati** [comment]
> Still ends up pretty complex [in the @overload(s)](https://github.com/pola-rs/polars/blob/ff994f0502bc5ce886df43ffbbf468a79adcf281/py-polars/polars/dataframe/frame.py#L1199-L1235), but much easier to differentiate the parts:

I agree, my plan was actually to add these in this or follow-up PR. I find them easier to read and follow

File: narwhals/dataframe.py (hunk 4 above)
Comment (dangotbanned) [reply narwhals/dataframe.py:872]: @EdAbati interesting 🤔

So should this line in the doc be rewritten?

https://github.com/narwhals-dev/narwhals/blob/733ab52873ee31c6e1eb2e76f2ef44df27b97e29/narwhals/dataframe.py#L928

I do appreciate that what you're saying is documented, but saying *always* reads to me as something that belongs in the annotation?

Comment (EdAbati) [reply narwhals/dataframe.py:872]: Mmm oh yeah actually good catch, you are right! then it makes a lot of sense

Comment (dangotbanned) [reply narwhals/dataframe.py:872]: @EdAbati marking as unresolved just to show another option if a reviewer wanted it:

```diff
diff --git a/narwhals/dataframe.py b/narwhals/dataframe.py
index 6b673a80..207ef7b2 100644
--- a/narwhals/dataframe.py
+++ b/narwhals/dataframe.py
@@ -859,6 +859,8 @@ class DataFrame(BaseFrame[DataFrameT]):
         """
         return self._compliant_frame.estimated_size(unit=unit)  # type: ignore[no-any-return]
 
+    @overload
+    def __getitem__(self: DataFrame[pd.DataFrame], item: int) -> Any: ...
     @overload
     def __getitem__(  # type: ignore[overload-overlap]
         self: Self,
diff --git a/narwhals/stable/v1/__init__.py b/narwhals/stable/v1/__init__.py
index 5aefefe2..af8aed47 100644
--- a/narwhals/stable/v1/__init__.py
+++ b/narwhals/stable/v1/__init__.py
@@ -84,6 +84,7 @@ if TYPE_CHECKING:
     from types import ModuleType
 
     import numpy as np
+    import pandas as pd
     from typing_extensions import Self
 
     from narwhals.dtypes import DType
@@ -131,6 +132,8 @@ class DataFrame(NwDataFrame[IntoDataFrameT]):
     def _lazyframe(self: Self) -> type[LazyFrame[Any]]:
         return LazyFrame
 
+    @overload
+    def __getitem__(self: DataFrame[pd.DataFrame], item: int) -> Any: ...
     @overload
     def __getitem__(  # type: ignore[overload-overlap]
         self: Self,

```

I think this describes the `pandas` edge case?

File: tests/frame/getitem_test.py [hunk 5]
```diff
@@ -55,15 +55,15 @@ def test_slice_lazy_fails() -> None:
...
+    result = nw.from_native(constructor_eager(data), eager_only=True)[1]
     assert_equal_data(result, {"a": [2], "b": [12]})
 
 
 def test_slice_fails(constructor_eager: ConstructorEager) -> None:
     class Foo: ...
 
     with pytest.raises(TypeError, match="Expected str or slice, got:"):
-        nw.from_native(constructor_eager(data), eager_only=True)[Foo()]  # type: ignore[call-overload]
+        nw.from_native(constructor_eager(data), eager_only=True)[Foo()]  # type: ignore[call-overload, unused-ignore]
```
Comment (MarcoGorelli) [inline tests/frame/getitem_test.py:66]: 🤔 ignoring unused-ignore seems a bit strange?

Comment (dangotbanned) [reply tests/frame/getitem_test.py:66]: @MarcoGorelli see https://github.com/narwhals-dev/narwhals/pull/1963#discussion_r1947720414

It is to ignore `pre-commit` not understanding the call is invalid

**Review by MarcoGorelli — APPROVED** [review]
thanks both!

File: tests/frame/getitem_test.py (hunk 5 above)
Comment (MarcoGorelli) [reply tests/frame/getitem_test.py:66]: thanks!
